File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 1
    bridge = deque([0] * (bridge_length-1))
    truck_que = deque(truck_weights)
    sum_weights = 0
    
    while truck_que:
        sum_weights += truck_que[0]
        if truck_que:
            if sum_weights <= weight:
                bridge.append(truck_que.popleft())
            else:
                sum_weights -= truck_que[0]
                bridge.append(0)
        sum_weights -= bridge.popleft()
        answer += 1
        
    return answer+len(bridge)

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Deque<Integer> truck = new ArrayDeque<>();
        Deque<Integer> bridge = new ArrayDeque<>(Collections.nCopies(bridge_length-1, 0));
        for (int t : truck_weights) {
            truck.offer(t);
        }
        
        int wei = 0;
        while (!truck.isEmpty()) {
            wei += truck.peek();
            if (wei > weight) {
                wei -= truck.peek();
                bridge.offer(0);
            } else {
                bridge.offer(truck.poll());
            }
            wei -= bridge.poll();
            answer++;
        }
        return answer+bridge.size()+1;
    }
}